### ISSUE TITLE
Explainability results from IntegratedGradients differ using PyG's to_captum function

### ISSUE THREAD

**Issue by jan0b** [pyg-team/pytorch_geometric#4708]
### 🐛 Describe the bug

# Description
I am working on an explainability task using the IntegratedGradient method und Captum's implementation. My code example adopts PyG's own explainability example, but uses the UPFD dataset and its example model by PyG for graph-classification.
In my understanding the resulting gradients for the edges and nodes (when calculated individually) should be the same as when calculating them together in one attribution. This means:
```
ig_attr_node1 == ig_attr_node2
ig_attr_edge1 == ig_attr_edge2
```
should be **True** which they aren't. I would apprecciate some help or clarification why different gradients would be desired, if I am wrong.

# Code
```
import os.path as osp

import matplotlib.pyplot as plt
import torch
import torch.nn.functional as F
from torch.nn import Linear
from captum.attr import IntegratedGradients

from torch_geometric.loader import DataLoader
import torch_geometric.transforms as T
from torch_geometric.datasets import Planetoid, GNNBenchmarkDataset, UPFD
from torch_geometric.nn import Explainer, GCNConv, to_captum, SAGEConv, global_max_pool

dataset = 'PATTERN'
path = "<>"
# dataset = Planetoid(path, dataset, transform=T.NormalizeFeatures())
dataset = UPFD(path, 'politifact', transform=T.NormalizeFeatures(), feature='profile')
train_loader = DataLoader(dataset, batch_size=128, shuffle=True)

data = dataset[0]

class Net(torch.nn.Module):
    def __init__(self, model, in_channels, hidden_channels, out_channels,
                 concat=False):
        super().__init__()
        self.concat = concat

        if model == 'GCN':
            self.conv1 = GCNConv(in_channels, hidden_channels)
        elif model == 'SAGE':
            self.conv1 = SAGEConv(in_channels, hidden_channels)
        elif model == 'GAT':
            self.conv1 = GATConv(in_channels, hidden_channels)


        if self.concat:
            self.lin0 = Linear(in_channels, hidden_channels)
            self.lin1 = Linear(2 * hidden_channels, hidden_channels)

        self.lin2 = Linear(hidden_channels, out_channels)

    def forward(self, x, edge_index, batch):
        h = self.conv1(x, edge_index).relu()
        h = global_max_pool(h, batch)

        if self.concat:
            # Get the root node (tweet) features of each graph:
            root = (batch[1:] - batch[:-1]).nonzero(as_tuple=False).view(-1)
            root = torch.cat([root.new_zeros(1), root + 1], dim=0)
            news = x[root]

            news = self.lin0(news).relu()
            h = self.lin1(torch.cat([news, h], dim=-1)).relu()

        h = self.lin2(h)
        return h.log_softmax(dim=-1)


device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net('GCN', dataset.num_features, 128,
            dataset.num_classes, concat=True).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=0.01)


def train():
    model.train()

    total_loss = 0
    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data.x, data.edge_index, data.batch)
        loss = F.nll_loss(out, data.y)
        loss.backward()
        optimizer.step()
        total_loss += float(loss) * data.num_graphs

    return total_loss / len(train_loader.dataset)

for epoch in range(1, 61):
    loss = train()
    # train_acc = test(train_loader)
    # val_acc = test(val_loader)
    # test_acc = test(test_loader)
    # print(f'Epoch: {epoch:02d}, Loss: {loss:.4f}, Train: {train_acc:.4f}, '
    #       f'Val: {val_acc:.4f}, Test: {test_acc:.4f}')
print("DONE!")

#======EXPLAINABILITY STARTS HERE======

output_idx = 0
target = int(data.y[output_idx])
batch = torch.zeros(data.x.shape[0], dtype=int)
model.to('cpu')
model.eval()

# Edge explainability
# ===================

# Captum assumes that for all given input tensors, dimension 0 is
# equal to the number of samples. Therefore, we use unsqueeze(0).
captum_model = to_captum(model, mask_type='edge', output_idx=output_idx)
edge_mask = torch.ones(data.num_edges, requires_grad=True) # device=device

ig = IntegratedGradients(captum_model)
ig_attr_edge1, delta = ig.attribute(edge_mask.unsqueeze(0), target=target,
                            baselines=torch.zeros(edge_mask.unsqueeze(0).shape),
                            additional_forward_args=(data.x, data.edge_index, batch),
                            internal_batch_size=1,
                            return_convergence_delta=True)

# ig_attr_edge1 = ig.attribute(edge_mask.unsqueeze(0), target=target,
#                             additional_forward_args=(data.x, data.edge_index, batch),
#                             internal_batch_size=1)

# Scale attributions to [0, 1]:
ig_attr_edge = ig_attr_edge1.squeeze(0).abs()
ig_attr_edge /= ig_attr_edge.max()

# Visualize absolute values of attributions:
explainer = Explainer(model)
ax, G = explainer.visualize_subgraph(node_idx=None, edge_index=data.edge_index, edge_mask=ig_attr_edge)
plt.show()

# Node explainability
# ===================

captum_model = to_captum(model, mask_type='node', output_idx=output_idx)

ig = IntegratedGradients(captum_model)
ig_attr_node1 = ig.attribute(data.x.unsqueeze(0), target=target,
                            additional_forward_args=(data.edge_index, batch),
                            internal_batch_size=1)

# Scale attributions to [0, 1]:
ig_attr_node = ig_attr_node1.squeeze(0).abs().sum(dim=1)
ig_attr_node /= ig_attr_node.max()

# Visualize absolute values of attributions:
ax, G = explainer.visualize_subgraph(node_idx=None, edge_index=data.edge_index, edge_mask=torch.ones(data.num_edges),
                                     node_alpha=ig_attr_node)
plt.show()

# Node and edge explainability
# ============================

captum_model = to_captum(model, mask_type='node_and_edge',
                         output_idx=output_idx)

ig = IntegratedGradients(captum_model)
ig_attr_node2, ig_attr_edge2 = ig.attribute(
    inputs=(data.x.unsqueeze(0), edge_mask.unsqueeze(0)), target=target,
    baselines=(torch.zeros(data.x.unsqueeze(0).shape), torch.zeros(edge_mask.unsqueeze(0).shape)),
    additional_forward_args=(data.edge_index, batch), internal_batch_size=1)

# Scale attributions to [0, 1]:
ig_attr_node = ig_attr_node2.squeeze(0).abs().sum(dim=1)
ig_attr_node /= ig_attr_node.max()
ig_attr_edge = ig_attr_edge2.squeeze(0).abs()
ig_attr_edge /= ig_attr_edge.max()

# Visualize absolute values of attributions:
ax, G = explainer.visualize_subgraph(None, data.edge_index, ig_attr_edge,
                                     node_alpha=ig_attr_node)
plt.show()
```

### Environment

* PyG version: 2.0.4
* PyTorch version: 1.11.0
* OS: Linux 4.18.0
* Python version: 3.8.13
* CUDA/cuDNN version: cuda11.3_cudnn8.2.0_0
* How you installed PyTorch and PyG (`conda`, `pip`, source): Pytorch and PyG with conda
* Any other relevant information (*e.g.*, version of `torch-scatter`):---

**Comment by jan0b** [comment]
Thank you for your quick answer and sorry for labeling my issue wrong. From the formula for integrated gradients (taken from the paper):

$$
IntegratedGrads_i(x) ::= (x_i-x_i') \times \int_{\alpha=0}^{1} \frac{\partial F(x'+\alpha\times(x-x'))}{\partial x_i}d\alpha
$$

I understand that the gradient from the baseline to the input is not dependent on any other variable than the input (and the baseline, but that is set to zero). Shouldn't then the integrated gradient for e.g. an edge be independent on whether the integrated gradient for a node is calculated?

**Comment by rusty1s** [comment]
In case you want to explain both nodes and edges simultaneously, they will both be considered as input. The gradients of both node features and edges are dependent on each other.
